File: demos/react-native/modules/offline/android/src/main/java/cc/musicme/offline/expo/PrefetchCache.kt
package cc.musicme.offline.expo

import cc.musicme.offline.ByteSource

object PrefetchCache {
    private val sources = HashMap<String, ByteSource>()

    @Synchronized
    fun put(trackId: String, source: ByteSource) {
        sources[trackId]?.close()
        sources[trackId] = source
    }

    @Synchronized
    fun take(trackId: String): ByteSource? = sources.remove(trackId)

    @Synchronized
    fun clear(except: String? = null) {
        sources.filterKeys { it != except }.values.forEach { it.close() }
        if (except != null) {
            val kept = sources[except]
            sources.clear()
            if (kept != null) sources[except] = kept
        } else {
            sources.clear()
        }
    }
}

File: demos/react-native/modules/offline/ios/PrefetchCache.swift
import Foundation

final class PrefetchCache: @unchecked Sendable {
    nonisolated(unsafe) static let shared = PrefetchCache()
    private let lock = NSLock()
    private var sources: [String: ByteSource] = [:]

    private init() {}

    func put(_ trackId: String, _ source: ByteSource) {
        lock.lock(); defer { lock.unlock() }
        sources[trackId]?.close()
        sources[trackId] = source
    }

    /// Returns + removes the cached source. Caller owns lifetime after this.
    func take(_ trackId: String) -> ByteSource? {
        lock.lock(); defer { lock.unlock() }
        return sources.removeValue(forKey: trackId)
    }

    func clear(except keepTrackId: String? = nil) {
        lock.lock(); defer { lock.unlock() }
        for (k, v) in sources where k != keepTrackId { v.close() }
        if let keep = keepTrackId, let kept = sources[keep] {
            sources = [keep: kept]
        } else {
            sources.removeAll()
        }
    }
}
